src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -338,7 +343,11 @@ void GcsNodeManager::AddNode(std::shared_ptr<rpc::GcsNodeInfo> node) {
     // Add an empty resources for this node.
     RAY_CHECK(cluster_resources_.emplace(node_id, rpc::ResourceMap()).second);
     // Register this node to the `node_failure_detector_` which will start monitoring it.
-    node_failure_detector_->AddNode(node_id);
+    // NOTE: TO AVOID HEARTBEAT BEING DEALYED BY MAIN THREAD, MAKE SURE NODE ADDING IS
+    // ALWAYS HANDLED BY ITS OWN IO SERVICE
```

Comment:
```suggestion
```suggestion
    // Note: To avoid heartbeats being delayed by main thread, make sure heartbeat is always
    // handled by its own IO service.
```
```